chain/chain/src/chain.rs

```diff
@@ -1206,6 +1277,63 @@ impl Chain {
         }
     }
 
+    /// Check if we can request chunks for this orphan. Conditions are
+    /// 1) Among the `NUM_ORPHAN_ANCESTORS_CHECK` immediate parents of the block at least one is
+    ///    accepted(is in store), call it `ancestor`
+    /// 2) The next block of `ancestor` has the same epoch_id as the orphan block
+    ///    (This is because when requesting chunks, we will use `ancestor` hash instead of the
+    ///     previous block hash of the orphan to decide epoch id
+    /// 3) All the immediate parents of the block after `ancestor` are either orphans or in
+    ///     `blocks_with_missing_chunks`
+    /// 4) Orphans that with outstanding missing chunks request has not exceed `MAX_ORPHAN_MISSING_CHUNKS`
+    /// 5) Orphans have missing chunks
+    pub fn should_request_chunks_for_orphan(
+        &mut self,
+        me: &Option<AccountId>,
+        orphan: &Block,
+    ) -> Option<OrphanMissingChunksInfo> {
+        if self.orphans_requested_missing_chunks.len() >= MAX_ORPHAN_MISSING_CHUNKS {
+            return None;
+        }
+        if self.orphans_requested_missing_chunks.contains(orphan.hash()) {
+            return None;
+        }
+        let mut block_hash = *orphan.header().prev_hash();
+        for _ in 1..NUM_ORPHAN_ANCESTORS_CHECK {
+            if self.get_block(&block_hash).is_ok() {
+                if let Ok(epoch_id) = self.runtime_adapter.get_epoch_id_from_prev_block(&block_hash)
+                {
+                    if &epoch_id == orphan.header().epoch_id() {
+                        let mut chain_update = self.chain_update();
+                        if let Err(e) = chain_update.ping_missing_chunks(me, block_hash, &orphan) {
+                            return match e.kind() {
+                                ErrorKind::ChunksMissing(missing_chunks) => {
+                                    Some(OrphanMissingChunksInfo {
+                                        missing_chunks,
+                                        epoch_id,
+                                        ancestor_hash: block_hash,
+                                    })
+                                }
+                                _ => None,
+                            };
+                        }
+                    }
+                }
+                return None;
```

Comment:
I looked at where this function is called. It seems to me that this part of the code may be called multiple times for a block if it is deep in the orphan pool, since we call it every time we call `check_orphan`. We may want to consider caching some intermediate results

Reply:
The function checks if this orphan is already in `orphans_requested_missing_chunks` and skip the rest of the checks if it is. This check is also only called for a new orphan or children of a block that is just accepted that are within depth 5. The only case where I think this function can be called multiple times on the an orphan is if the orphan does not have missing chunks when it was first added, which is actually impossible now, since a node can't process chunks for orphans.